pyrate/core/phase_closure/sum_closure.py

```diff
@@ -88,23 +91,30 @@ def sum_phase_closures(ifg_files: List[str], loops: List[WeightedLoop], params:
         # for k, r in enumerate(rets):
         #     closure_dict[k], ifgs_breach_count_dict[k] = r
         # TODO: enable multiprocessing - needs pickle error fix
-        ifgs_breach_count_arr = []
+        ifgs_breach_count = np.zeros(shape=(ifgs[0].phase_data.shape + (n_ifgs,)), dtype=np.uint16)
         for k, weighted_loop in enumerate(loops):
             closure_dict[k], ifgs_breach_count_l = __compute_ifgs_breach_count(weighted_loop, edge_to_indexed_ifgs,
                                                                                params)
-            ifgs_breach_count_arr.append(ifgs_breach_count_l)
-        ifgs_breach_count = np.sum(np.stack(ifgs_breach_count_arr), axis=0)
+            ifgs_breach_count += ifgs_breach_count_l
     else:
         loops_with_index = list(enumerate(loops))
         process_loops = mpiops.array_split(loops_with_index)
-        ifgs_breach_count_arr = []
+        ifgs_breach_count = np.zeros(shape=(ifgs[0].phase_data.shape + (n_ifgs,)), dtype=np.uint16)
         for k, weighted_loop in process_loops:
             closure_dict[k], ifgs_breach_count_l = __compute_ifgs_breach_count(weighted_loop, edge_to_indexed_ifgs,
                                                                                params)
-            ifgs_breach_count_arr.append(ifgs_breach_count_l)
+            ifgs_breach_count += ifgs_breach_count_l  # process
         closure_dict = join_dicts(mpiops.comm.gather(closure_dict, root=0))
-        ifgs_breach_count_process = np.sum(np.stack(ifgs_breach_count_arr), axis=0)
-        ifgs_breach_count = mpiops.comm.reduce(ifgs_breach_count_process, op=mpiops.sum0_op, root=0)
+
+        total_gb = mpiops.comm.allreduce(ifgs_breach_count.nbytes / 1e9, op=mpiops.MPI.SUM)
+        log.info("Memory usage due to ifgs_breach_count {:2.4f}GB of data".format(total_gb))
+        log.debug(f"shape of ifgs_breach_count is {ifgs_breach_count.shape}")
+        log.debug(f"dtype of ifgs_breach_count is {ifgs_breach_count.dtype}")
+       
+        mpiops.comm.Reduce(ifgs_breach_count, ifgs_breach_count, op=MPI.SUM, root=0) 
```

Comment:
I pushed a commit which changes this to a capital R reduce. This fixes further issues that lower-case reduce was having with creating a 13 GB pickle object to communicate its part of the `ifgs_breach_count_arr` back to the root node. Using `Reduce` is safe because while `mpiops.array_split` may result in different numbers of loops being sent to each process, the accumulated breach count arrays are all numpy arrays of the same size `(height of ifg, width of ifg, # of ifgs)` so we can use preallocated buffers for our reduction operation (and the sum over axis 0 is default behaviour of `MPI.SUM` so we don't need to mess with a custom `sum0_op`). Still a couple of things to change before we can merge this:
1. I'm not sure if it's safe to use the same buffer for sending and receiving for the reduction - if not we'll need to create an empty array of zeros
2. This commit will break the non-MPI option, because I had to directly import mpi4py in `sum_closure.py` to get `MPI.SUM`. Should be easy to fix this by exporting MPI.SUM out of mpiops.

Reply:
good pick up on the `Reduce` use here Richard. Thank you.